subsys/bluetooth/mesh/pb_gatt_srv.c

```diff
@@ -115,14 +115,12 @@ static void gatt_disconnected(struct bt_conn *conn, uint8_t reason)
 
 	bt_conn_get_info(conn, &info);
 	if (info.role != BT_CONN_ROLE_PERIPHERAL || !service_registered ||
-	    info.id != BT_ID_DEFAULT) {
+	    info.id != BT_ID_DEFAULT || !cli || cli->conn != conn) {
```

Comment:
I don't want to say, the existing logic is getting harder and harder to maintain later, and refactoring needs to be considered

Reply:
Not sure how this part can be fixed. `connected` and `disconnected` callbacks are too generic.